### PR TITLE
Add query authorization 

### PR DESCRIPTION
**Pull request by dain**
This is now based on #2755 as having an identity in the web UI is required for query access to work, so please ignore the first commit "Add simple login page to UI".

### CODE REVIEW

**Comment by tooptoop4** [comment]
https://github.com/prestodb/presto/issues/13300 https://github.com/prestodb/presto/issues/13297

File: presto-main/src/main/java/io/prestosql/server/QueryResource.java [hunk 1]
```diff
@@ -68,55 +83,76 @@ public QueryResource(DispatchManager dispatchManager)
...
+
+        try {
+            BasicQueryInfo queryInfo = dispatchManager.getQueryInfo(queryId);
+            canKillQueryOwnedBy(extractAuthorizedIdentity(servletRequest, accessControl), queryInfo.getSession().getUser(), accessControl);
+            dispatchManager.cancelQuery(queryId);
+        }
+        catch (AccessDeniedException e) {
+            throw new ForbiddenException();
+        }
+        catch (NoSuchElementException ignored) {
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/server/QueryResource.java]: It wasn't catched before. Why do you need to handle this now? If it was a bug, please extract this to separate commit.

File: presto-main/src/test/java/io/prestosql/server/TestQueryResource.java [hunk 2]
```diff
@@ -111,6 +144,49 @@ public void testGetQueryInfoExecutionFailure()
         assertEquals(info.getFailureInfo().getErrorCode(), DIVISION_BY_ZERO.toErrorCode());
     }
 
+    @Test
+    public void testCancel()
+    {
+        String queryId = startQuery("SELECT * FROM tpch.sf100.lineitem");
```
Comment (kokosing) [inline presto-main/src/test/java/io/prestosql/server/TestQueryResource.java]: You could introduce a testing procedure or function to sleep instead of wasting CPU.

File: presto-main/src/test/java/io/prestosql/server/TestQueryResource.java [hunk 3]
```diff
@@ -111,6 +144,49 @@ public void testGetQueryInfoExecutionFailure()
...
+        testKilled("killed");
+    }
+
+    @Test
+    public void testPreempted()
+    {
+        testKilled("preempted");
+    }
+
+    private void testKilled(String preempted)
```
Comment (kokosing) [inline presto-main/src/test/java/io/prestosql/server/TestQueryResource.java]: s/preempted/message? or reason?

File: presto-main/src/test/java/io/prestosql/server/TestQueryResource.java [hunk 4]
```diff
@@ -111,6 +144,49 @@ public void testGetQueryInfoExecutionFailure()
...
+
+        server.getAccessControl().deny(privilege("query", KILL_QUERY));
+        try {
+            assertEquals(cancelQueryInfo(queryId), 403);
+        }
+        finally {
+            server.getAccessControl().reset();
+        }
+
+        assertEquals(cancelQueryInfo(queryId), 204);
```
Comment (kokosing) [inline presto-main/src/test/java/io/prestosql/server/TestQueryResource.java]: can you run query `system.runtime.queries` to see that query was cancelled

File: presto-main/src/test/java/io/prestosql/server/TestQueryResource.java [hunk 5]
```diff
@@ -111,6 +144,49 @@ public void testGetQueryInfoExecutionFailure()
...
+
+        server.getAccessControl().deny(privilege("query", KILL_QUERY));
+        try {
+            assertEquals(killQueryInfo(queryId, preempted), 403);
+        }
+        finally {
+            server.getAccessControl().reset();
+        }
+
+        assertEquals(killQueryInfo(queryId, preempted), 202);
```
Comment (kokosing) [inline presto-main/src/test/java/io/prestosql/server/TestQueryResource.java]: can you run query `system.runtime.queries` to see that query was killed and possibly check the reason why query was killed

File: presto-main/src/test/java/io/prestosql/server/TestQueryStateInfoResource.java [hunk 6]
```diff
@@ -133,6 +136,38 @@ public void testGetQueryStateInfo()
         assertNotNull(info);
     }
 
+    @Test
+    public void testGetAllQueryStateInfosDenied()
```
Comment (kokosing) [inline presto-main/src/test/java/io/prestosql/server/TestQueryStateInfoResource.java]: can you please add tests were some queries are filtered and some are not

File: presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/security/FileBasedSystemAccessControl.java [hunk 7]
```diff
@@ -179,6 +179,27 @@ public void checkCanSetUser(Optional<Principal> principal, String userName)
...
+
+    @Override
+    public void canViewQueryOwnedBy(Optional<SystemSecurityContext> context, String queryOwner)
+    {
+    }
+
+    @Override
+    public Set<String> filterViewQueryOwnedBy(Optional<SystemSecurityContext> context, Set<String> queryOwner)
+    {
+        return queryOwner;
```
Comment (kokosing) [inline presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/security/FileBasedSystemAccessControl.java]: queryOwner -> queryOwners

here and everywhere else

File: presto-testing/src/main/java/io/prestosql/testing/AbstractTestQueries.java [hunk 8]
```diff
@@ -4688,6 +4689,7 @@ public void testFilterPushdownWithAggregation()
     @Test
     public void testAccessControl()
     {
+        assertAccessAllowed("SELECT * FROM orders", privilege("orders", EXECUTE_QUERY));
```
Comment (kokosing) [inline presto-testing/src/main/java/io/prestosql/testing/AbstractTestQueries.java]: do you need similar test with `assertAccessDenied`?

File: presto-tests/src/test/java/io/prestosql/connector/system/TestKillQuery.java [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+        executor.shutdownNow();
+    }
+
+    @Test(timeOut = 60_000)
+    public void testKillQuery()
+            throws Exception
+    {
+        String testQueryId = "test_query_id_" + ThreadLocalRandom.current().nextInt(100_000_000);
+        Future<?> queryFuture = executor.submit(() -> {
+            getQueryRunner().execute(format("SELECT count(comment) as %s FROM tpch.sf100000.lineitem", testQueryId));
```
Comment (kokosing) [inline presto-tests/src/test/java/io/prestosql/connector/system/TestKillQuery.java]: It would be nice to use `sleep` UDF

File: presto-tests/src/test/java/io/prestosql/connector/system/TestKillQuery.java [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+    @AfterClass(alwaysRun = true)
+    public void tearDown()
+    {
+        executor.shutdownNow();
+    }
+
+    @Test(timeOut = 60_000)
+    public void testKillQuery()
```
Comment (kokosing) [inline presto-tests/src/test/java/io/prestosql/connector/system/TestKillQuery.java]: Have you seen `io.prestosql.connector.system.TestSystemConnector`?

**Comment by dain** [comment]
I am updating this for the latest authentication changes.

File: presto-main/src/test/java/io/prestosql/server/TestQueryResource.java (hunk 2 above)
Comment (dain) [reply presto-main/src/test/java/io/prestosql/server/TestQueryResource.java]: This technique is used by several tests (I copied it from somewhere else).  I agree they should all be update, but that seems like an orthogonal project.

File: presto-docs/src/main/sphinx/security/built-in-system-access-control.rst [hunk 11]
```diff
@@ -143,6 +144,30 @@ and deny all other access, you can use the following rules:
...
+-----------
+
+These rules control the ability of a user to execute, view, or kill a query. The user is
+granted or denied access, based on the first matching rule read from top to bottom. If no
+rules are specified, no checks are performed. If no rule matches, query management is not
+allowed. Each rule is composed of the following fields:
+
+* ``user`` (optional): regex to match against user name. Defaults to ``.*``.
+* ``owner`` (optional): regex to match against the query owner name. Defaults to ``.*``.
+* ``allow`` (required): set of query permissions granted to user.
```
Comment (electrum) [inline presto-docs/src/main/sphinx/security/built-in-system-access-control.rst]:
```
* ``allow`` (required): set of query permissions granted to user. Values: ``execute``, ``view``, ``kill``

File: presto-product-tests/src/main/java/io/prestosql/tests/TestSqlCancel.java [hunk 12]
```diff
@@ -168,7 +168,9 @@ public Response cancel(String queryId)
         {
             requireNonNull(queryId, "queryId is null");
             URI cancelUri = uriBuilderFrom(uri).appendPath("/v1/query").appendPath(queryId).build();
-            Request request = prepareDelete().setUri(cancelUri).build();
+            Request request = prepareDelete()
+                    .setHeader("X-Presto-User", "anyUser")
+                    .setUri(cancelUri).build();
```
Comment (electrum) [inline presto-product-tests/src/main/java/io/prestosql/tests/TestSqlCancel.java]: `build` on next line

File: presto-spi/src/main/java/io/prestosql/spi/security/SystemAccessControl.java [hunk 13]
```diff
@@ -75,6 +78,47 @@ default void checkCanSetUser(Optional<Principal> principal, String userName)
         denySetUser(principal, userName);
     }
 
+    /**
+     * Checks if identity can execute a query.
+     *
+     * @throws AccessDeniedException if not allowed
+     */
+    default void canExecuteQuery(SystemSecurityContext context)
```
Comment (electrum) [inline presto-spi/src/main/java/io/prestosql/spi/security/SystemAccessControl.java]: These should all be `checkCanXxx`

File: presto-main/src/main/java/io/prestosql/security/AccessControlUtil.java [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+    public static List<BasicQueryInfo> filterQueries(Identity identity, List<BasicQueryInfo> queries, AccessControl accessControl)
+    {
+        String currentUser = identity.getUser();
+        Set<String> owners = queries.stream()
+                .map(BasicQueryInfo::getSession)
+                .map(SessionRepresentation::getUser)
+                .filter(owner -> !owner.equals(currentUser))
+                .collect(toImmutableSet());
+        owners = accessControl.filterQueriesOwnedBy(identity, owners);
+        owners = Sets.union(owners, ImmutableSet.of(currentUser));
```
Comment (electrum) [inline presto-main/src/main/java/io/prestosql/security/AccessControlUtil.java]: This is a view. Replace with
```java
owners = ImmutableSet.builder()
    .add(currentUser)
    .addAll(owners)
    .build();

File: presto-main/src/test/java/io/prestosql/server/TestQueryResource.java [hunk 15]
```diff
@@ -89,6 +106,17 @@ public void testGetQueryInfos()
...
+
+        server.getAccessControl().deny(privilege("query", VIEW_QUERY));
+        try {
+            assertTrue(getQueryInfos("/v1/query").isEmpty());
+            assertTrue(getQueryInfos("/v1/query?state=finished").isEmpty());
+            assertTrue(getQueryInfos("/v1/query?state=failed").isEmpty());
+            assertTrue(getQueryInfos("/v1/query?state=running").isEmpty());
+        }
+        finally {
+            server.getAccessControl().reset();
```
Comment (kokosing) [inline presto-main/src/test/java/io/prestosql/server/TestQueryResource.java]: It would be nice to have some closeable entity.

```
try (AutoCloseasble privilege = deny(privilege("query", VIEW_QUERY)) { 
   ...
}

...
private AutoCloseable deny(... privilege) {
   server.getAccessControl().deny(privilege);
   return () -> server.getAccessControl().reset();
}
```

File: presto-main/src/test/java/io/prestosql/server/TestQueryResource.java [hunk 16]
```diff
@@ -99,6 +127,16 @@ public void testGetQueryInfoDispatchFailure()
         assertFalse(info.isScheduled());
         assertNotNull(info.getFailureInfo());
         assertEquals(info.getFailureInfo().getErrorCode(), SYNTAX_ERROR.toErrorCode());
+
+        server.getAccessControl().deny(privilege("query", VIEW_QUERY));
+        try {
+            assertThatThrownBy(() -> getQueryInfo(queryId))
```
Comment (kokosing) [inline presto-main/src/test/java/io/prestosql/server/TestQueryResource.java]: extract `assertThatForbidden(...)`?

File: presto-main/src/test/java/io/prestosql/server/TestQueryResource.java [hunk 17]
```diff
@@ -111,6 +149,60 @@ public void testGetQueryInfoExecutionFailure()
...
+
+        server.getAccessControl().deny(privilege("query", KILL_QUERY));
+        try {
+            assertEquals(cancelQueryInfo(queryId), 403);
+        }
+        finally {
+            server.getAccessControl().reset();
+        }
+
+        assertEquals(cancelQueryInfo(queryId), 204);
```
Comment (kokosing) [inline presto-main/src/test/java/io/prestosql/server/TestQueryResource.java]: call this method twice to see how it behaves when user cancel cancelled (or killed) query. Unless it is tested somewhere else.

File: presto-main/src/test/java/io/prestosql/server/TestQueryResource.java [hunk 18]
```diff
@@ -111,6 +149,60 @@ public void testGetQueryInfoExecutionFailure()
...
+
+        server.getAccessControl().deny(privilege("query", KILL_QUERY));
+        try {
+            assertEquals(killQueryInfo(queryId, killType), 403);
+        }
+        finally {
+            server.getAccessControl().reset();
+        }
+
+        assertEquals(killQueryInfo(queryId, killType), 202);
```
Comment (kokosing) [inline presto-main/src/test/java/io/prestosql/server/TestQueryResource.java]: kill it twice

File: presto-main/src/test/java/io/prestosql/server/TestQueryResource.java [hunk 19]
```diff
@@ -111,6 +149,60 @@ public void testGetQueryInfoExecutionFailure()
...
+        testKilled("killed");
+    }
+
+    @Test
+    public void testPreempted()
+    {
+        testKilled("preempted");
+    }
+
+    private void testKilled(String killType)
```
Comment (kokosing) [inline presto-main/src/test/java/io/prestosql/server/TestQueryResource.java]: Introduce an enum

File: presto-main/src/test/java/io/prestosql/server/TestQueryStateInfoResource.java [hunk 20]
```diff
@@ -127,12 +142,67 @@ public void testGetQueryStateInfosForUserNoResult()
...
+                    prepareGet()
+                            .setUri(server.resolve("/v1/queryState"))
+                            .setHeader(PRESTO_USER, executionUser)
+                            .build(),
+                    createJsonResponseHandler(listJsonCodec(QueryStateInfo.class)));
+
+            assertEquals(infos.size(), expectedCount);
+        }
+        finally {
+            server.getAccessControl().reset();
```
Comment (kokosing) [inline presto-main/src/test/java/io/prestosql/server/TestQueryStateInfoResource.java]: same comment about `AutoCloseable`

File: presto-main/src/test/java/io/prestosql/server/TestQueryStateInfoResource.java [hunk 21]
```diff
@@ -127,12 +142,67 @@ public void testGetQueryStateInfosForUserNoResult()
...
+        }
+    }
+
+    @Test
+    public void testGetQueryStateInfoDenied()
+    {
+        server.getAccessControl().deny(privilege("query", VIEW_QUERY));
```
Comment (kokosing) [inline presto-main/src/test/java/io/prestosql/server/TestQueryStateInfoResource.java]: when you change the `server` state that suggests that you cannot run tests in paralel. Notice that we used to run such tests within `io.prestosql.tests.AbstractTestQueryFramework#executeExclusively`

File: presto-tests/src/test/java/io/prestosql/connector/system/TestKillQuery.java [hunk 22]
```diff
@@ -0,0 +1,111 @@
...
+            throws Exception
+    {
+        String testQueryId = "test_query_id_" + ThreadLocalRandom.current().nextInt(100_000_000);
+        Future<?> queryFuture = executor.submit(() -> {
+            getQueryRunner().execute(format("SELECT count(comment) as %s FROM tpch.sf100000.lineitem", testQueryId));
+        });
+
+        Optional<Object> queryIdValue = Optional.empty();
+        while (!queryIdValue.isPresent()) {
+            Thread.sleep(50);
```
Comment (kokosing) [inline presto-tests/src/test/java/io/prestosql/connector/system/TestKillQuery.java]: use failsafe?

File: presto-tests/src/test/java/io/prestosql/connector/system/TestKillQuery.java [hunk 23]
```diff
@@ -0,0 +1,111 @@
...
+                    .collect(toOptional());
+        }
+        String queryId = queryIdValue.get().toString();
+
+        assertFalse(queryFuture.isDone());
+
+        getQueryRunner().getAccessControl().deny(privilege("query", KILL_QUERY));
+        assertThatThrownBy(() -> getQueryRunner().execute(getSession("other_user"), format("CALL system.runtime.kill_query('%s', 'should fail')", queryId)))
+                .hasMessageContaining("Cannot kill query");
+        getQueryRunner().getAccessControl().reset();
```
Comment (kokosing) [inline presto-tests/src/test/java/io/prestosql/connector/system/TestKillQuery.java]: shouldn't this be called within `finally`?

File: presto-spi/src/main/java/io/prestosql/spi/security/AccessDeniedException.java [hunk 24]
```diff
@@ -51,6 +51,36 @@ public static void denySetUser(Optional<Principal> principal, String userName, S
...
+    }
+
+    public static void denyViewQuery()
+    {
+        denyViewQuery(null);
+    }
+
+    public static void denyViewQuery(String extraInfo)
+    {
+        throw new AccessDeniedException(format("Cannot view query%s", formatExtraInfo(extraInfo)));
```
Comment (kokosing) [inline presto-spi/src/main/java/io/prestosql/spi/security/AccessDeniedException.java]: I don't see a test that would check that this exception is thrown

File: presto-docs/src/main/sphinx/security/built-in-system-access-control.rst [hunk 25]
```diff
@@ -144,6 +145,31 @@ and deny all other access, you can use the following rules:
...
     }
 
+.. _query_rules:
+
+Query Rules
+-----------
+
+These rules control the ability of a user to execute, view, or kill a query. The user is
+granted or denied access, based on the first matching rule read from top to bottom. If no
+rules are specified, no checks are performed. If no rule matches, query management is not
```
Comment (kokosing) [inline presto-docs/src/main/sphinx/security/built-in-system-access-control.rst]: I would add to `, no checks are performed` that all access is allowed. It might be not clear what `no checks` mean.

File: presto-docs/src/main/sphinx/security/built-in-system-access-control.rst [hunk 26]
```diff
@@ -144,6 +145,31 @@ and deny all other access, you can use the following rules:
...
+allowed. Each rule is composed of the following fields:
```
Comment (kokosing) [inline presto-docs/src/main/sphinx/security/built-in-system-access-control.rst]: `not allowed` -> `denied`?

File: presto-docs/src/main/sphinx/security/query-access.json [hunk 27]
```diff
@@ -0,0 +1,20 @@
...
+      "allow": true
+    }
+  ],
+  "queries": [
+    {
+      "user": "admin",
+      "allow": ["execute", "kill", "view"]
+    },
+    {
+      "user": "alice",
```
Comment (kokosing) [inline presto-docs/src/main/sphinx/security/query-access.json]: I would replace `alice` with `user_than_can_execute_or_kill` same for `allow` user. Also it looks like you are missing a typical case where user is only allowed to `view`.

File: presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/security/QueryAccessRule.java [hunk 28]
```diff
@@ -0,0 +1,89 @@
...
+import java.util.Optional;
+import java.util.Set;
+import java.util.regex.Pattern;
+
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static java.util.Arrays.stream;
+import static java.util.Objects.requireNonNull;
+import static java.util.function.Function.identity;
+
+public class QueryAccessRule
```
Comment (kokosing) [inline presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/security/QueryAccessRule.java]: add `toString`

File: presto-main/src/main/java/io/prestosql/security/AccessControl.java [hunk 29]
```diff
@@ -46,6 +46,35 @@
...
+     */
+    void checkCanExecuteQuery(Identity identity);
+
+    /**
+     * Checks if identity can view a query owned by the specified user.  The method
+     * will not be called when the current user is the query owner.
+     *
+     * @throws AccessDeniedException if not allowed
+     */
+    void checkCanViewQueryOwnedBy(Identity identity, String queryOwner);
```
[the remark this answers is not in the document]
Comment (dain) [reply presto-main/src/main/java/io/prestosql/security/AccessControl.java]: This is the pattern of the access control checks.  It is possible to implement one in terms of the other, but is not clear which direction you would go.  In some security systems one call is significantly cheaper than the other, which is why we have both.

File: presto-main/src/main/java/io/prestosql/server/HttpRequestSessionContext.java [hunk 30]
```diff
@@ -172,6 +176,32 @@ else if (nameParts.size() == 2) {
...
 
+    public static Identity extractAuthorizedIdentity(HttpServletRequest servletRequest, HttpHeaders httpHeaders, AccessControl accessControl)
+    {
+        return extractAuthorizedIdentity(
+                Optional.ofNullable((Identity) servletRequest.getAttribute(AUTHENTICATED_IDENTITY)),
+                httpHeaders.getRequestHeaders(),
+                accessControl);
+    }
+
+    public static Identity extractAuthorizedIdentity(Optional<Identity> optionalAuthenticatedIdentity, MultivaluedMap<String, String> headers, AccessControl accessControl)
```
[the remark this answers is not in the document]
Comment (dain) [reply presto-main/src/main/java/io/prestosql/server/HttpRequestSessionContext.java]: It is that way because the code below reuses that name.  I spent a lot of time playing with this code and that is the best naming convention I could come up with.

File: presto-tests/src/test/java/io/prestosql/connector/system/TestKillQuery.java (hunk 22 above)
Comment (dain) [reply presto-tests/src/test/java/io/prestosql/connector/system/TestKillQuery.java]: This test already has a timeout, so I don't think we need anything more.

File: presto-main/src/main/java/io/prestosql/security/AccessControl.java (hunk 29 above)
[the remark this answers is not in the document]
Comment (dain) [reply presto-main/src/main/java/io/prestosql/security/AccessControl.java]: I don't think we should make that change now.  I don't think we have enough information to design this for the future (e.g., how is it going to work when we implement proper SQL roles).  For the specific use case, the implementer of such a system could simply pull the groups again, which they may want to do anyway, since the group membership would be old.

**Comment by tooptoop4** [comment]
@dain does hiding queries run from other users work on v1/query endpoint, system.runtime.queries table, jmx tables/endpoints in addition to standard web ui ?

**Comment by dain** [comment]
@tooptoop4 it should.  There are tests for all of the above in this PR.  If I missed something, please file an issue.
